### PR TITLE
Removed the OpenSSL usage

### PR DESCRIPTION
**Pull request by ezimuel**
This PR removes the usage of the `openssl_random_pseudo_bytes()` function of OpenSSL due to the bug [#70014](https://bugs.php.net/bug.php?id=70014) and the discussion reported [here](https://github.com/paragonie/random_compat/issues/96).

### CODE REVIEW

File: doc/book/zend.math.introduction.md [hunk 1]
```diff
@@ -12,20 +12,13 @@ We expect to add more functionalities in the future.
...
-function of the OpenSSL extension if installed. If the OpenSSL extension is not present in the
-system the algorithm tries to use the the
-[mcrypt\_create\_iv](http://it.php.net/manual/en/function.mcrypt-create-iv.php) function of the
-Mcrypt extension (using the `MCRYPT_DEV_URANDOM` parameter). Finally, if the OpenSSL and Mcrypt are
-not installed the generator uses the `mt_rand` function of PHP.
+uses different approaches. If PHP 7 is running we used the cryptographically secure pseudo-random
+functions [random_bytes](http://php.net/manual/en/function.random-bytes.php) and
+[random_int](http://php.net/manual/en/function.random-int.php), otherwise we use the
+[Mcrypt](http://it.php.net/manual/en/book.mcrypt.php) extension or /dev/urandom source with a mixer
+function provided by . If you don't have a secure random source in your environment the component
```
Comment (weierophinney) [inline doc/book/zend.math.introduction.md]: "provided by "... what?

File: src/Rand.php [hunk 2]
```diff
@@ -39,11 +40,8 @@ public static function getBytes($length, $strong = false)
             return false;
         }
 
-        if (function_exists('openssl_random_pseudo_bytes')) {
-            $bytes = openssl_random_pseudo_bytes($length, $usable);
-            if (true === $usable) {
-                return $bytes;
-            }
+        if (version_compare(PHP_VERSION, '7.0.0') >= 0) {
```
Comment (Ocramius) [inline src/Rand.php]: `function_exists('random_bytes')` would be preferable here

File: src/Rand.php [hunk 3]
```diff
@@ -117,6 +114,9 @@ public static function getInteger($min, $max, $strong = false)
                 'The min parameter must be lower than max parameter'
             );
         }
+        if (version_compare(PHP_VERSION, '7.0.0') >= 0) {
```
Comment (Ocramius) [inline src/Rand.php]: `function_exists('random_int')`

**Comment by ezimuel** [comment]
@Ocramius I used function_exists() as suggested, thanks!

**Comment by ezimuel** [comment]
@weierophinney I think the PR is ready to be merged.

File: src/Rand.php [hunk 4]
```diff
@@ -56,7 +54,7 @@ public static function getBytes($length, $strong = false)
         if (true === $strong && false === $checkAlternatives) {
             throw new Exception\RuntimeException(
                 'This PHP environment doesn\'t support secure random number generation. ' .
-                'Please consider installing the OpenSSL and/or Mcrypt extensions'
+                'Please consider installing Mcrypt extension or use PHP 7'
```
Comment (weierophinney) [inline src/Rand.php]: Rephrase this to "Please consider either installing ext/mcrypt or upgrading to PHP 7".

File: src/Rand.php [hunk 5]
```diff
@@ -76,8 +74,7 @@ public static function getAlternativeGenerator()
         if (!class_exists('RandomLib\\Factory')) {
             throw new Exception\RuntimeException(
                 'The RandomLib fallback pseudorandom number generator (PRNG) '
-                . ' must be installed in the absence of the OpenSSL and '
-                . 'Mcrypt extensions'
+                . ' must be installed in the absence of a secure source'
```
Comment (weierophinney) [inline src/Rand.php]: This is a bit cryptic; as an end-user, if I don't know cryptography, I might not understand what "absence of a secure source" refers to. Is it cryptic on purpose? i.e., to prevent an attacker from understanding the capabilities of the server if the exception is displayed?

**Comment by weierophinney** [comment]
@ezimuel Just a few notes about exceptions to address, and then I can merge. I'm assuming this would be for a new minor release?

File: src/Rand.php (hunk 5 above)
Comment (ezimuel) [inline src/Rand.php]: @weierophinney this is not cryptic by purpose, but definitely you find out a good point here. We can rephrase with "The RandomLib fallback pseudorandom generator is not installed. Please install it to support secure random numbers". What do you think?

Comment (weierophinney) [inline src/Rand.php]: Perfect.

**Comment by weierophinney** [comment]
Merged to develop for release with 2.7.0.

**Comment by j-schumann** [comment]
Isn't this considered backwards incompatible because developer action is required when you happen to have neither mcrypt nor the RandomLib installed?

After updating to 2.7.0 my application now throws the "The RandomLib fallback pseudorandom generator is not installed" exception when using Zend\Validator\Csrf (which calls Zend\Math\Rand::getBytes).

Shouldn't have the openSSL usage been marked as deprecated before in 2.6.0 so developers could prepare? I understand that there are security implications but are they important enough to force the BC break?

**Comment by Ocramius** [comment]
If this is considered a security issue, it takes priority over BC compat. Depends on @ezimuel's decision tho.

**Comment by gianarb** [comment]
>  Depends on @ezimuel's decision tho.

Is it a security issue @ezimuel ? :) Thanks

**Comment by j-schumann** [comment]
I see, thanks for clearing that up!
One could argue that, since #70014 is patched for all PHP versions from 5.4.44 on, the security risk is not sincere.
I just would opt that this could have been fixed without a BC break, e.g. if Zend\Math\Rand would implement the functionality of RandomLib\Source\URandom and RandomLib\Source\CAPICOM itself as it already checks for the availability of this sources already. A new major version (or after deprecation) could then remove the functionality and require RandomLib or ext/mcrypt.

**Comment by ezimuel** [comment]
@j-schumann yes, this is considered as a potential security issue for some specific OpenSSL installation, see [#70014](https://bugs.php.net/bug.php?id=70014).

**Comment by gianarb** [comment]
Thanks! :)

**Comment by ezimuel** [comment]
@j-schumann there are also other issues around the usage of OpenSSL, see comments [reported here](https://github.com/paragonie/random_compat/issues/96).
